mod-source-record-manager-server/src/main/java/org/folio/services/flowcontrol/RawRecordsFlowControlServiceImpl.java

```diff
@@ -85,80 +84,99 @@ public void trackChunkReceivedEvent(String tenantId, Integer recordsCount) {
+  public void trackRecordCompleteEvent(String tenantId, Integer recordsCount) {
     if (!enableFlowControl) {
       return;
     }
 
-    currentState.merge(tenantId, -recordsCount, Integer::sum);
-    decreaseCounterInDb(tenantId, recordsCount);
-
-    LOGGER.info("--------------- Tenant: [{}]. Chunk duplicate event comes, update current value to: {} ---------------", tenantId, recordsCount);
-
-    resumeIfThresholdAllows(tenantId);
+    decreaseState(tenantId, recordsCount);
+    LOGGER.debug("trackRecordCompleteEvent:: Tenant: [{}]. Record count: {}, Current state:{}",
+      tenantId, recordsCount, currentState.get(tenantId));
   }
 
   @Override
-  public void trackRecordCompleteEvent(String tenantId, Integer actualCounterValue) {
+  public void trackChunkDuplicateEvent(String tenantId, Integer recordsCount) {
     if (!enableFlowControl) {
       return;
     }
 
-    if (actualCounterValue < 0) {
-      LOGGER.info("trackRecordCompleteEvent:: Tenant: [{}]. Current value less than zero, because of single record imports or resetting state, back to zero...", tenantId);
-      currentState.put(tenantId, 0);
-      eventProcessedService.resetEventsToProcess(tenantId);
-    } else {
-      currentState.put(tenantId, actualCounterValue);
-    }
-
-    LOGGER.info("--------------- Tenant: [{}]. Current value after complete event: {} ---------------", tenantId, actualCounterValue);
+    decreaseState(tenantId, recordsCount);
+    LOGGER.debug("trackChunkDuplicateEvent:: Tenant: [{}]. Record count: {}, Current state:{}",
+      tenantId, recordsCount, currentState.get(tenantId));
+  }
 
+  private void decreaseState(String tenantId, Integer recordsCount) {
+    initFetchMode(tenantId);
+    decreaseCounterInDb(tenantId, recordsCount);
     resumeIfThresholdAllows(tenantId);
   }
 
   private void resumeIfThresholdAllows(String tenantId) {
-    if (currentState.get(tenantId) <= recordsThreshold) {
-      Collection<KafkaConsumerWrapper<String, String>> rawRecordsReadConsumers = consumersStorage.getConsumersByEvent(DI_RAW_RECORDS_CHUNK_READ.value());
 
-      rawRecordsReadConsumers.forEach(consumer -> {
-        if (consumer.demand() == 0) {
-          consumer.resume();
+    LOGGER.info("resumeIfThresholdAllows :: Tenant: [{}]. Current state:{}, History state: {}",
+      tenantId, currentState.get(tenantId), historyState.get(tenantId));
 
-          LOGGER.info("resumeIfThresholdAllows:: Tenant: [{}]. Kafka consumer - id: {}, subscription - {} is resumed for all tenants, because {} met threshold {}",
-            tenantId, consumer.getId(), DI_RAW_RECORDS_CHUNK_READ.value(), this.currentState, recordsThreshold);
+    historyState.putIfAbsent(tenantId, Pair.of(0,0));
+    if (historyState.get(tenantId).getKey().equals(currentState.get(tenantId))) {
+      Integer previousValue = historyState.get(tenantId).getValue();
+      historyState.put(tenantId, Pair.of(historyState.get(tenantId).getKey(), ++previousValue));
```

Comment:
I think you need to use `computeIfPresent` instead of getting a previous value and then incrementing it. This action should happen atomically, getting and incrementing by one.

Reply:
Ready. Thanks